transport_analysis/viscosity.py

```diff
@@ -0,0 +1,215 @@
+        keys = {
+            "x": [0],
+            "y": [1],
+            "z": [2],
+            "xy": [0, 1],
+            "xz": [0, 2],
+            "yz": [1, 2],
+            "xyz": [0, 1, 2],
+        }
+
+        try:
+            _dim = keys[dim_str]
+        except KeyError:
+            raise ValueError(
+                "invalid dim_type: {} specified, please specify one of xyz, "
+                "xy, xz, yz, x, y, z".format(dim_str)
+            )
+
+        return _dim, len(_dim)
+
+    def _single_frame(self):
+        """
+        Constructs arrays of velocities and positions
+        for viscosity computation.
+        """
+        # This runs once for each frame of the trajectory
+
+        # The trajectory positions update automatically
+        # You can access the frame number using self._frame_index
+
+        # trajectory must have velocity and position information
+        if not (
+            self._ts.has_velocities
+            and self._ts.has_positions
+            and self._ts.volume != 0
+        ):
+            raise NoDataError(
+                "Helfand viscosity computation requires "
+                "velocities, positions, and box volume in the trajectory"
+            )
+
+        # fill volume array
+        self._volumes[self._frame_index] = self._ts.volume
+
+        # set shape of velocity array
+        self._velocities[self._frame_index] = self.atomgroup.velocities[
+            :, self._dim
+        ]
+
+        # set shape of position array
+        self._positions[self._frame_index] = self.atomgroup.positions[
+            :, self._dim
+        ]
+
+    def _conclude(self):
+        """
+        Calculates the viscosity function via the simple "windowed" algorithm.
+        """
+        self._vol_avg = np.average(self._volumes)
+
+        lagtimes = np.arange(1, self.n_frames)
+
+        # iterate through all possible lagtimes from 1 to number of frames
+        for lag in lagtimes:
+            # get difference of momentum * position shifted by "lag" frames
+            diff = (
+                self._masses_rs
+                * self._velocities[:-lag, :, :]
+                * self._positions[:-lag, :, :]
+                - self._masses_rs
+                * self._velocities[lag:, :, :]
+                * self._positions[lag:, :, :]
+            )
+
+            # square and sum each x(, y, z) diff per particle
+            sq_diff = np.square(diff).sum(axis=-1)
+
+            # average over # frames
+            # update viscosity by particle array
+            self.results.visc_by_particle[lag, :] = np.mean(sq_diff, axis=0)
+
+        # divide by 2, boltzmann constant, vol_avg, and temp_avg
+        self.results.visc_by_particle = self.results.visc_by_particle / (
+            2 * constants[self.boltzmann] * self._vol_avg * self.temp_avg
+        )
+        # average over # particles and update results array
+        self.results.timeseries = self.results.visc_by_particle.mean(axis=1)
```

Comment:
Add plotting?

Reply:
I wanted to do that in a separate PR to avoid making this PR too bloated. Would it be better to add it here?

Reply:
thats fine to do next PR